Implements #7921 extend query to support latest versions

HermesSbicego-Laser opened the pull request:
Implements #7921, branched from #7920 cause we need the LatestValue column to be there.
When a query is created/edited it is possible to specify the Content Version to query (Published or Latest).
Based on that choice the HqlQuery will filter 
- the right VersionOption 
- the right column of the `(*)IndexFilterRecord`

Based on that choice the HqlQuery will order
- using the right column of the `(*)IndexFilterRecord`

HermesSbicego-Laser committed:
- Adds the column LatestValue to all type of FieldIndexRecord
- Adds Indexes to increase performances in queries
- Adds Handler to store LatestValue during Update event
- Adds Service to Get and Set the LatestValue of a field
- Adds heavy projections index tables update to the Upgrade module

HermesSbicego-Laser committed:
- Adds a new Transaction foreach content update in order to prevent tables locks during import

HermesSbicego-Laser committed:
- Adds Index over FieldIndexPartRecord_Id

HermesSbicego-Laser committed:
- Adds Column ScopeVersion to QueryPartRecord and manage it during the part's lifecycle
- Manage Queries using ScopeVersion informations

HermesSbicego-Laser committed:
- Manage the scope of the query within Fields

HermesSbicego-Laser committed:
- Manage the sort criterion for fields based on the VersionScope of the query

sebastienros commented on src/Orchard.Web/Modules/Orchard.Projections/Migrations.cs:
```diff
@@ -282,5 +281,38 @@ public class Migrations : DataMigrationImpl {
+
+            return 5;
+        }
+        public int UpdateFrom5() {
```
Move to `UpdateFrom4`. And update the `Create` method to create the table correctly including all subsequent changes, then return `5`.

HermesSbicego-Laser replied:
Ok, I will threat this PR and the #7920 one as they were the same PR, so I can aggregate UpdateFrom4 and 5

sebastienros commented on src/Orchard.Web/Modules/Orchard.Projections/Migrations.cs:
```diff
@@ -282,5 +281,38 @@ public class Migrations : DataMigrationImpl {
+            SchemaBuilder.AlterTable("DecimalFieldIndexRecord", table => table
+            .AddColumn<decimal>("LatestValue"));
+
+            //Adds indexes for better performances in queries
```
Two comments:
- We need to talk about the "necessity" of these indices
- You are reusing the same index name multiple times, is it an issue? should it be unique per database, or no otherwise table prefix would break it?

HermesSbicego-Laser replied:
- Necessity of these indexes: do you mean SQL indexes or the "LatestValue" columns? For SQL indexes, I think at least `IX_FieldIndexPartRecord_Id` should be created because it is a foreign key "in fact". The index over PropertyName column is for projections and HqlQueries so I think they should be there. I don't see a good reason why they shouldn't be there
- Same index name: I seem to remember that Indexes name should be unique per table and not per database. At least in SQL Server, so i don't see an issue here. Perhaps it's not the same in MySql or Postgre?

sebastienros commented on src/Orchard.Web/Modules/Orchard.Projections/Providers/Filters/ContentFieldsFilter.cs:
```diff
@@ -75,7 +75,9 @@ public class ContentFieldsFilter : IFilterProvider {
 
             // generate the predicate based on the editor which has been used
-            Action<IHqlExpressionFactory> predicate = fieldTypeEditor.GetFilterPredicate(context.State);
+            dynamic fullState = context.State;
```
Why `dynamic` ? You can't access the `VersionScope` property directly?

HermesSbicego-Laser replied:
fieldTypeEditor.GetFilterPredicate (2 lines above) method use a dynamic parameter as input. I need that information within IFieldTypeEditor.GetFilterPredicate in order to return the right column (Value or LatestValue) to query.

sebastienros commented on src/Orchard.Web/Modules/Orchard.Projections/Providers/SortCriteria/ContentFieldsSortCriteria.cs:
```diff
@@ -79,11 +79,11 @@ public class ContentFieldsSortCriterion : ISortCriterionProvider {
-                ? context.Query.OrderBy(relationship, x => x.Asc("Value")) 
-                : context.Query.OrderBy(relationship, x => x.Desc("Value"));
+            context.Query = ascending
+                ? context.Query.OrderBy(relationship, x => x.Asc(context.VersionScope == QueryVersionScopeOptions.Latest ? "LatestValue" : "Value"))
```
Maybe create a method in the context to get the field name (based on the scope). I have seen this multiple times.

HermesSbicego-Laser replied:
Ok, i will try

sebastienros commented on src/Orchard.Web/Modules/Orchard.Projections/Services/ProjectionManager.cs:
```diff
@@ -168,7 +168,8 @@ public class ProjectionManager : IProjectionManagerExtension {
                     Query = groupQuery,
-                    State = FormParametersHelper.ToDynamic(sortCriterion.State)
+                    State = FormParametersHelper.ToDynamic(sortCriterion.State),
+                    VersionScope = queryRecord.VersionScope
```
Instead of duplicating this property in all context objects, can't we just pass the parent Query which already  has the value?

HermesSbicego-Laser replied:
Do you mean to add the QueryPart to the context?

sebastienros commented on src/Orchard.Web/Modules/Orchard.Projections/Views/EditorTemplates/Parts/QueryPart_Edit.cshtml:
```diff
@@ -0,0 +1,8 @@
+@using Orchard.Projections;
+
+<fieldset>
+    @Html.LabelFor(m => m.VersionScope, T("Query Version Scope"))
```
"Version Scope" is not very intuitive. "Versions" only seems better. What do we use in Lucene indices?

HermesSbicego-Laser replied:
Ok for "Versions", or "Content versions"?

sebastienros commented on src/Orchard.Web/Modules/Upgrade/Controllers/ProjectionsController.cs:
```diff
@@ -0,0 +1,84 @@
+        }
+
+        [HttpPost, ActionName("Index")]
+        public ActionResult IndexPOST() {
```
Shouldn't it do it by batch of content items? If there are too many of them this will fail. There are some examples doing it in other upgrade controllers.

HermesSbicego-Laser replied:
Can you address me to an example where I can copy from? thanks.

HermesSbicego-Laser replied:
Found it by myself in InfosetController.cs

sebastienros commented on src/Orchard.Web/Modules/Orchard.Projections/Handlers/FieldIndexPartHandler.cs:
```diff
@@ -43,24 +46,56 @@ public class FieldIndexPartHandler : ContentHandler {
         }
+        private void Updated(UpdateContentContext context, FieldIndexPart fieldIndexPart) {
+            if (context.UpdatingItemVersionRecord.Latest) { // updates projection draft indexes only if it is the latest version
+                foreach (var part in fieldIndexPart.ContentItem.Parts) {
```
Try to externalize this method, and use a lambda as the differences.

HermesSbicego-Laser replied:
OK, I will try...

sebastienros commented on src/Orchard.Web/Modules/Orchard.Projections/Services/DraftFieldIndexService.cs:
```diff
@@ -0,0 +1,125 @@
+using Orchard.Projections.Models;
+
+namespace Orchard.Projections.Services {
+    public class DraftFieldIndexService : IDraftFieldIndexService {
```
I understand this is to prevent a breaking change. But I'd prefer to see it merged into the existing service in the dev branch. Can you file an issue for this? I will assign it for 1.11.

HermesSbicego-Laser replied:
Ok, i will do it.

HermesSbicego-Laser replied:
opened issue #7925

HermesSbicego-Laser committed:
- Adds the source QueryPartRecord which generates the IHqlQuery to FilterContext and SortCriteriaContext: used in ProjectionManager
- Adds a method which returns the columnName to Sort/Filter to FilterContext and SortCriteriaContext: used in ContentFieldsSortCriteria
- Externalize the logic to calculate the column to index
- Merge Update 4 and 5
- Modified the label "Query Version Scope" into "Content's version"

HermesSbicego-Laser committed:
- Batch upgrade for LatestValue

HermesSbicego-Laser committed:
- async ajax call in order to prevent freezing UI

HermesSbicego-Laser committed:
- Created a context for lambda diversification

BenedekFarkas commented:
Does the code expect that the Query's version scope can only be Published or All? I'm asking because #8111.

HermesSbicego-Laser commented:
In this PR, the Query's version scope  can be Published or Latest.

HermesSbicego-Laser commented:
@BenedekFarkas I read a bit the code in your #8112 PR. Adding the filter on the right versionOption of the content is not sufficient. Should be changed other code. In particular where we test the VersionOption in order to choose the right column (Value/LatestValue) to check/order.
see:  
1. https://github.com/OrchardCMS/Orchard/pull/7923/files#diff-2e5f3c3cb7233e791056952d3b891828
2. https://github.com/OrchardCMS/Orchard/pull/7923/files#diff-718cbbbafe30601793f25355e00903f1
3. classes implementing IFieldTypeEditor

BenedekFarkas commented:
@HermesSbicego-Laser good point, I only saw the addition of LatestValue after submitting my PR - I'll look into it and update accordingly, thanks!
